File: test/src/population/statistics.cpp
/*
    Genesis - A toolkit for working with phylogenetic data.
    Copyright (C) 2014-2020

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    Contact:
    Lucas Czech <lucas.czech@h-its.org>
    Exelixis Lab, Heidelberg Institute for Theoretical Studies
    Schloss-Wolfsbrunnenweg 35, D-69118 Heidelberg, Germany
*/

/**
 * @brief
 *
 * @file
 * @ingroup test
 */

#include "src/common.hpp"

#include "genesis/population/functions/statistics.hpp"

using namespace genesis::population;
using namespace genesis::utils;

TEST( Population, Statistics )
{
    // Run loop for some lower values.
    for( size_t p = 1; p < 50; ++p ) {
        for( size_t c = 1; c < 3 * p; ++c ) {
            EXPECT_FLOAT_EQ( n_base_matrix( c, p ), n_base( c, p ));
        }
    }

    // Also test some higher values. We request the higher coverage values first in order to
    // avoid recomputing the matrix as much as possible.
    EXPECT_FLOAT_EQ( n_base_matrix(  500,  100 ), n_base(  500,  100 ));
    EXPECT_FLOAT_EQ( n_base_matrix(  100,  100 ), n_base(  100,  100 ));
    EXPECT_FLOAT_EQ( n_base_matrix(   50,  100 ), n_base(   50,  100 ));
    EXPECT_FLOAT_EQ( n_base_matrix( 2000, 1000 ), n_base( 2000, 1000 ));
    EXPECT_FLOAT_EQ( n_base_matrix(  500, 1000 ), n_base(  500, 1000 ));
    EXPECT_FLOAT_EQ( n_base_matrix(  100, 1000 ), n_base(  100, 1000 ));
}

File: python/test/util/math/statistics.py
# Genesis - A toolkit for working with phylogenetic data.
# Copyright (C) 2023-2025 Giannis Reppas and Lucas Czech
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Contact:
# Lucas Czech <lucas.czech@sund.ku.dk>
# University of Copenhagen, Globe Institute, Section for GeoGenetics
# Oster Voldgade 5-7, 1350 Copenhagen K, Denmark

import unittest
import tempfile
import os
import pathlib
import io

from genesis.genesis.util import math

class TestGenesisutilMathStatistics(unittest.TestCase):
    def test_arithmetic_mean(self):
        self.assertAlmostEqual( 0.0,  math.arithmetic_mean([ ]))
        self.assertAlmostEqual( 7.0,  math.arithmetic_mean([ 4.0, 6.0, 11.0 ]))

    def test_weighted_arithmetic_mean(self):
        self.assertAlmostEqual(  0.0,  math.weighted_arithmetic_mean([], []) )
        self.assertAlmostEqual( 86.0,  math.weighted_arithmetic_mean([ 80.0, 90.0 ], [ 20.0, 30.0 ]) )

    def test_geometric_mean(self):
        self.assertAlmostEqual( 0.0,  math.geometric_mean([ ]))
        self.assertAlmostEqual( 4.0,  math.geometric_mean([ 2.0, 8.0 ]))

        self.assertRaises(
            ValueError,
            math.geometric_mean,
            [ 0.0 ]
        )
        self.assertRaises(
            ValueError,
            math.geometric_mean,
            [ -1.0 ]
        )

    def test_weighted_geometric_mean(self):
        self.assertAlmostEqual( 0.0,  math.weighted_geometric_mean([], []) )
        self.assertAlmostEqual( 1.0,  math.weighted_geometric_mean([ 1.0 ], [ 1.0 ]) )
        self.assertAlmostEqual( 6.0,  math.weighted_geometric_mean([ 6.0, 6.0 ], [ 3.0, 8.0 ]) )

        self.assertRaises(
            ValueError,
            math.weighted_geometric_mean,
            [  0.0 ], [ 1.0 ]
        )
        self.assertRaises(
            ValueError,
            math.weighted_geometric_mean,
            [ -1.0 ], [ 1.0 ]
        )
        self.assertRaises(
            RuntimeError,
            math.weighted_geometric_mean,
            [ 5.0 ], [ 1.0, 2.0 ]
        )
        self.assertRaises(
            RuntimeError,
            math.weighted_geometric_mean,
            [ 5.0, 2.0 ], [ 1.0 ]
        )

    def test_harmonic_mean(self):
        self.assertAlmostEqual(
            2.0, math.harmonic_mean([ 1.0, 4.0, 4.0 ],      math.HarmonicMeanZeroPolicy.kIgnore)
        )
        self.assertAlmostEqual(
            2.0, math.harmonic_mean([ 1.0, 4.0, 4.0, 0.0 ], math.HarmonicMeanZeroPolicy.kIgnore)
        )

    def test_weighted_harmonic_mean(self):
        self.assertAlmostEqual( 2.0,  math.weighted_harmonic_mean(
        [ 1.0, 4.0, 4.0 ], [ 1.0, 1.0, 1.0 ],
            math.HarmonicMeanZeroPolicy.kIgnore
        ))
        self.assertAlmostEqual( 2.0,  math.weighted_harmonic_mean(
            [ 1.0, 4.0, 4.0, 0.0 ], [ 1.0, 1.0, 1.0, 1.0 ],
            math.HarmonicMeanZeroPolicy.kIgnore
        ))

        self.assertRaises(
            ValueError,
            math.weighted_harmonic_mean,
            [  0.0 ], [ 1.0 ]
        )
        self.assertRaises(
            ValueError,
            math.weighted_harmonic_mean,
            [ -1.0 ], [ 1.0 ]
        )
        self.assertRaises(
            RuntimeError,
            math.weighted_harmonic_mean,
            [ 5.0 ], [ 1.0, 2.0 ]
        )
        self.assertRaises(
            RuntimeError,
            math.weighted_harmonic_mean,
            [ 5.0, 2.0 ], [ 1.0 ]
        )

    # def test_norms(self):
    #     vals = [ 1.0, 2.0, 3.0 ]
    #     self.assertAlmostEqual( 6.0,  math.manhattan_norm(vals))
    #     self.assertAlmostEqual( 3.0,  math.maximum_norm(vals))
    #     self.assertAlmostEqual( 3.7416573867739413,   math.euclidean_norm(vals))
    #     self.assertAlmostEqual( 0.78566403520078676,  math.aitchison_norm(vals))

    # def test_median(self):
    #     self.assertAlmostEqual( 1.5,  math.median([ 0, 1, 2, 3 ]))
